src/config.rs

```diff
@@ -85,23 +131,87 @@ impl LayeredConfigs {
 
     /// Creates new merged config.
     pub fn merge(&self) -> config::Config {
-        let config_sources = [
-            Some(&self.default),
-            Some(&self.env_base),
-            self.user.as_ref(),
-            self.repo.as_ref(),
-            Some(&self.env_overrides),
-            self.arg_overrides.as_ref(),
-        ];
-        config_sources
+        self.sources()
             .into_iter()
-            .flatten()
+            .map(|(_, config)| config)
             .fold(config::Config::builder(), |builder, source| {
                 builder.add_source(source.clone())
             })
             .build()
             .expect("loaded configs should be merged without error")
     }
+
+    fn sources(&self) -> Vec<(ConfigSource, &config::Config)> {
+        let config_sources = [
+            (ConfigSource::Default, Some(&self.default)),
+            (ConfigSource::Env, Some(&self.env_base)),
+            (ConfigSource::User, self.user.as_ref()),
+            (ConfigSource::Repo, self.repo.as_ref()),
+            (ConfigSource::Env, Some(&self.env_overrides)),
+            (ConfigSource::CommandArg, self.arg_overrides.as_ref()),
+        ];
+        config_sources
+            .into_iter()
+            .filter_map(|(source, config)| config.map(|c| (source, c)))
+            .collect_vec()
+    }
+
+    pub fn resolved_config_values(&self) -> Result<Vec<AnnotatedValue>, ConfigError> {
+        // Collect annotated values from each config.
+        let mut configs = vec![];
+        fn annotate(value: config::Value, source: &ConfigSource) -> AnnotatedValue {
+            match value.kind {
+                config::ValueKind::Table(table) => AnnotatedValue::Table(
+                    table
+                        .into_iter()
+                        .map(|(k, v)| (k, annotate(v, source)))
+                        .collect(),
+                ),
+                _ => AnnotatedValue::Value {
+                    value,
+                    source: source.to_owned(),
+                    is_overridden: false,
+                },
+            }
+        }
+        for (source, config) in self.sources() {
+            configs.push(annotate(config.collect()?.into(), &source));
```

Comment:
We can process subtree of the given `name`. Or did you intentionally make `name` match any subtree key? 

Maybe I said `resolved_config_values(name: Option<&str>)` does too much stuff before, but I was wrong. It makes sense to narrow the config tables to process here.

Reply:
Yup, I like having some separation of concerns but it's a shame to materialize the entire config structure when most of it will be ignored to do it. I'll see if I can find a happy medium.

I was imagining some kind of generator/coroutine structure to decouple different aspects of the iteration more elegantly, but my rust-fu isn't strong enough for that yet, even if it would be a reasonable approach. 😅